src/container/srv_container.c

```diff
@@ -3440,6 +3513,168 @@ ds_cont_list(uuid_t pool_uuid, struct daos_pool_cont_info **conts,
 	return rc;
 }
 
+/* argument type for callback function to upgrade containers */
+struct upgrade_cont_iter_args {
+	uuid_t				 pool_uuid;
+
+	struct cont_svc			*svc;
+	struct rdb_tx			*tx;
+	/* total number of containers */
+	uint32_t			cont_nrs;
+	/* number of upgraded containers */
+	uint32_t			cont_upgraded_nrs;
+};
+
+/* callback function for upgrading containers iteration. */
+static int
+upgrade_cont_cb(daos_handle_t ih, d_iov_t *key, d_iov_t *val, void *varg)
+{
+	struct upgrade_cont_iter_args	*ap = varg;
+	uuid_t				 cont_uuid;
+	struct cont			*cont;
+	d_iov_t				 value;
+	uint64_t			 pda;
+	int				 rc;
+	bool				 upgraded = false;
+	uint32_t			 global_ver;
+	(void)val;
+
+	if (key->iov_len != sizeof(uuid_t)) {
+		D_ERROR("invalid key size: key="DF_U64"\n", key->iov_len);
+		return -DER_IO;
+	}
+
+	uuid_copy(cont_uuid, key->iov_buf);
+	D_DEBUG(DF_DSMS, "pool/cont: "DF_CONTF"\n",
+		DP_CONT(ap->pool_uuid, cont_uuid));
+
+	rc = cont_lookup(ap->tx, ap->svc, cont_uuid, &cont);
+	if (rc != 0) {
+		D_ERROR(DF_CONT": lookup cont failed, "DF_RC"\n",
+			DP_CONT(ap->pool_uuid, cont_uuid), DP_RC(rc));
+		return rc;
+	}
+
+	d_iov_set(&value, &global_ver, sizeof(global_ver));
+	rc = rdb_tx_lookup(ap->tx, &cont->c_prop,
+			   &ds_cont_prop_cont_global_version, &value);
+	if (rc && rc != -DER_NONEXIST)
+		goto out;
+	/* latest container, nothing to update */
+	if (rc == 0 && global_ver == DS_POOL_GLOBAL_VERSION)
+		goto out;
+
+	if (global_ver > DS_POOL_GLOBAL_VERSION) {
+		D_ERROR("Downgrading pool/cont: "DF_CONTF" not supported\n",
+			DP_CONT(ap->pool_uuid, cont_uuid));
+		rc = -DER_INVAL;
```

Comment:
Might be worth adding a new DER error for this case (`DER_NOTSUPPORTED` maybe?)

Reply:
Added.